backend/src/org/commcare/resources/model/ResourceTable.java

```diff
@@ -617,27 +695,47 @@ public boolean upgradeTable(ResourceTable incoming) throws UnresolvedResourceExc
     }
 
     /**
-     * Flag unstaged resources and those not present in replacement table for
-     * deletion.
+     * Uninstall table by removing unstaged resources and those not present in
+     * replacement table
+     *
+     * This method is the final step in an update, after this table has
+     * already been moved to a placeholder table and been evaluated for
+     * what resources are no longer necessary.
      *
-     * @param replacement Resources not in this table, flag for deletion
+     * If this table encounters any problems it will not intentionally
+     * throw errors, assuming that it's preferable to leave data unremoved
+     * rather than breaking the app.
+     *
+     * @param replacement Reference table; uninstall resources not also present
+     *                    in this table
      */
-    public void flagForDeletions(ResourceTable replacement) {
-        Stack<Resource> s = this.getResourceStack();
-        while (!s.isEmpty()) {
-            Resource r = s.pop();
-
-            if (replacement.getResourceWithId(r.getResourceId()) == null) {
-                // no entry in 'replacement' so it's no longer relevant
-                this.commit(r, Resource.RESOURCE_STATUS_DELETE);
-                continue;
-            }
-
-            if (r.getStatus() == Resource.RESOURCE_STATUS_UNSTAGED) {
-                // resource has been replaced, so flag for deletion
-                this.commit(r, Resource.RESOURCE_STATUS_DELETE);
+    public void uninstall(ResourceTable replacement) {
+        cleanup();
+        Hashtable<String, Resource> replacementMap = getResourceMap(replacement);
+        for (IStorageIterator it = storage.iterate(); it.hasMore(); ) {
+            Resource r = (Resource)it.nextRecord();
+            if (replacementMap.get(r.getResourceId()) == null ||
+                    r.getStatus() == Resource.RESOURCE_STATUS_UNSTAGED) {
+                // No entry in 'replacement' so it's no longer relevant
+                // OR resource has been replaced, so flag for deletion
+                try {
+                    r.getInstaller().uninstall(r);
+                } catch (Exception e) {
+                    Logger.log("Resource", "Error uninstalling resource " +
+                            r.getRecordGuid() + ". " + e.getMessage());
+                }
+            } else if (r.getStatus() == Resource.RESOURCE_STATUS_DELETE) {
+                // NOTE: Shouldn't be a way for this condition to occur, but check anyways...
+                try {
+                    r.getInstaller().uninstall(r);
+                } catch (Exception e) {
+                    Logger.log("Resource", "Error uninstalling resource " +
+                            r.getRecordGuid() + ". " + e.getMessage());
+                }
```

Comment:
Were there any places where the temp table "Clearing" could be "picked up" again in a state where the 

tempTable.completeUninstall();

Was called, but the flag for delete step was not? If not, I think you're right that we've already accepted the upgrade, so I can't find a way in my mind where this would need to be a two-step process